modules/jdbc/src/main/java/org/testcontainers/jdbc/ConnectionUrl.java

```diff
@@ -0,0 +1,200 @@
+     */
+    private String dbHostString;
+
+    private boolean inDaemonMode = false;
+
+    private Optional<String> databaseHost = Optional.empty();
+
+    private Optional<Integer> databasePort = Optional.empty();
+
+    private Optional<String> databaseName = Optional.empty();
+
+    private Optional<String> initScriptPath = Optional.empty();
+
+    private Optional<InitFunctionDef> initFunction = Optional.empty();
+
+    private Optional<String> queryString;
+
+    private Map<String, String> containerParameters;
+
+    private Map<String, String> queryParameters;
+
+    public static ConnectionUrl newInstance(final String url){
+        ConnectionUrl connectionUrl = new ConnectionUrl(url);
+        connectionUrl.parseUrl();
+        return connectionUrl;
+    }
+
+    private ConnectionUrl(final String url) {
+        this.url = Objects.requireNonNull(url, "Connection URL cannot be null");
+    }
+
+    public static boolean accepts(final String url) {
+        return url.startsWith("jdbc:tc:");
+    }
+
+    /**
+     * This method applies various REGEX Patterns to parse the URL associated with this instance.
+     * This is called from a @{@link ConnectionUrl#newInstance(String)} static factory method to create immutable instance of {@link ConnectionUrl}.
+     * To avoid mutation after class is instantiated, this method should not be publicly accessible.
+     */
+    private void parseUrl() {
+        /*
+        Extract from the JDBC connection URL:
+         * The database type (e.g. mysql, postgresql, ...)
+         * The docker tag, if provided.
+         * The URL query string, if provided
+       */
+        Matcher urlMatcher = Patterns.URL_MATCHING_PATTERN.matcher(this.getUrl());
+        if (!urlMatcher.matches()) {
+            //Try for Oracle pattern
+            urlMatcher = Patterns.ORACLE_URL_MATCHING_PATTERN.matcher(this.getUrl());
+            if (!urlMatcher.matches()) {
+                throw new IllegalArgumentException("JDBC URL matches jdbc:tc: prefix but the database or tag name could not be identified");
+            }
+        }
+        databaseType = urlMatcher.group(1);
+
+        imageTag = Optional.ofNullable(urlMatcher.group(3)).orElse("latest");
+
+        //String like hostname:port/database name, which may vary based on target database.
+        //Clients can further parse it as needed.
+        dbHostString = urlMatcher.group(4);
+
+        //In case it matches to the default pattern
+        Matcher dbInstanceMatcher = Patterns.DB_INSTANCE_MATCHING_PATTERN.matcher(dbHostString);
+        if (dbInstanceMatcher.matches()) {
+            databaseHost = Optional.of(dbInstanceMatcher.group(1));
+            databasePort = Optional.ofNullable(dbInstanceMatcher.group(3)).map(value -> Integer.valueOf(value));
+            databaseName = Optional.of(dbInstanceMatcher.group(4));
+        }
+
+        queryParameters = Collections.unmodifiableMap(
```

Comment:
this doesnt allow for duplicate params which might be very handy if for example I want to set jdbc://...?TC_INITSCRIPT=schema.sql&TC_INITSCRIPT=data.sql

Reply:
I see what you mean but I don't think it is possible even with the existing implementation. I would rather think of that as a separate feature request, maybe @rnorth can weigh in.

Reply:
Please let's see if we can tackle this without regular expressions separately. I'd be very happy if we could do that and make this code easier to read.

Reply:
Just making sure I understand it correctly. Do You mean, 'Allowing multiple TC_INITSCRIPT like parameter support' should be implemented in separate PR?

Reply:
On other thought, maybe a comma delimited list of files is accepted for single TC_INITSCRIPT parameter and executed in the sequence they are provided. Eg. `TC_INITSCRIPT=Script1.sql,Script2.sql,Script3.sql`.